### PR TITLE
Hisilicon: add hikey support

### PR DESCRIPTION
**Pull request by hzhuang1**
Fixe https://github.com/ARM-software/tf-issues/issues/421

Support BL1/BL2/BL2U/BL31 on HiKey platform. Support eMMC, UART,
GPIO, PMUSSI, fastboot, usb on HiKey platform. It could boot
system & flush images on HiKey.

In recovery mode (J1 pin5-6 are connected), BL1 & BL2U are loaded
into SRAM directly. So BL1 could boot BL2U to flush images.

In normal boot mode, only BL1 is loaded. Then it boots BL2.

Signed-off-by: Haojian Zhuang haojian.zhuang@linaro.org

### CODE REVIEW

**Comment by hzhuang1** [comment]
Since GPT, fastboot & usb driver are not merged, this pull request is just for code review.

**Comment by davidcunado-arm** [comment]
This PR is quite large - over 5000 loc. Please consider breaking into seperate logical patches as this will make it easier to review.

**Comment by hzhuang1** [comment]
OK

File: plat/hisilicon/hikey/hikey_def.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+					 HIKEY_BL1_MMC_DESC_SIZE)
+#define HIKEY_BL1_MMC_DATA_SIZE		0x0000B000
+
+#define MMC_BASE			0
+#define HIKEY_FIP_BASE			(MMC_BASE + (4 << 20))
+#define HIKEY_FIP_MAX_SIZE		(8 << 20)
+#define HIKEY_NSBL1U_BASE		(MMC_BASE + 0)
+#define HIKEY_NSBL1U_MAX_SIZE		(128 << 10)
+#define HIKEY_BL2U_BASE			(MMC_BASE + 0)
+#define HIKEY_BL2U_MAX_SIZE		(4 << 30)
```
Comment (danh-arm) [inline plat/hisilicon/hikey/hikey_def.h]: So it looks like you're storing BL2U in a separate part of MMC to the updateable FIP. This is good but could do with some explanation somewhere.

File: plat/hisilicon/hikey/hikey_io_storage.c [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+static const io_block_spec_t emmc_fip_spec = {
+	.offset		= HIKEY_FIP_BASE,
+	.length		= HIKEY_FIP_MAX_SIZE,
+};
+
+static const io_block_spec_t emmc_rpmb_spec = {
+	.offset		= HIKEY_NSBL1U_BASE,
+	.length		= HIKEY_NSBL1U_MAX_SIZE,
+};
```
Comment (danh-arm) [inline plat/hisilicon/hikey/hikey_io_storage.c]: You have references to NS_BL1U but I don't think you have one of these, right?

File: plat/hisilicon/hikey/include/platform_def.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+#define MAX_IO_DEVICES			3
+#define MAX_IO_HANDLES			4
+/* eMMC RPMB and eMMC User Data */
+#define MAX_IO_BLOCK_DEVICES		2
+
+/* GIC related constants (no GICR in GIC-400) */
+#define PLAT_ARM_GICD_BASE		0xF6801000
+#define PLAT_ARM_GICC_BASE		0xF6802000
+#define PLAT_ARM_GICH_BASE		0xF6804000
+#define PLAT_ARM_GICV_BASE		0xF6806000
```
Comment (danh-arm) [inline plat/hisilicon/hikey/include/platform_def.h]: Better to use your own names than these PLAT_ARM_ names and move to hikey_def.h.

File: plat/hisilicon/hikey/hikey_ddr.c [hunk 4]
```diff
@@ -0,0 +1,991 @@
...
+enum {
+	DDR_FREQ_533M = 0,
+	DDR_FREQ_800M,
+};
+
+static void init_pll(void)
+{
+	unsigned int data;
+
+	data = mmio_read_32((0xf7032000 + 0x000));
```
Comment (hzhuang1) [inline plat/hisilicon/hikey/hikey_ddr.c]: Hisilicon guys don't want to make DDR registers public. If those registers are defined as hardcoded number, they could accept.

File: plat/hisilicon/hikey/aarch64/hikey_common.c [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+					MT_DEVICE | MT_RW | MT_SECURE)
+
+/*
+ * Table of regions for different BL stages to map using the MMU.
+ * This doesn't include Trusted RAM as the 'mem_layout' argument passed to
+ * hikey_init_mmu_elx() will give the available subset of that,
+ */
+#if IMAGE_BL1
+static const mmap_region_t hikey_mmap[] = {
+	MAP_DDR,
```
Comment (hzhuang1) [inline plat/hisilicon/hikey/aarch64/hikey_common.c]: I forgot to remove it. At first, DDR was used in BL1. Now DDR is initialized in later stage. Only SRAM is used in BL1.

**Comment by gitfineon** [comment]
Hi @hzhuang1 ,

I red your message about this is just for review, but is it possible to add the files which are needed to compile a fip for hikey? It would be ok if not all features are ready, but compilable without them.

I tried to port the following files from [96boards-repo](https://github.com/96boards/arm-trusted-firmware)
```
new file:   drivers/synopsys/emmc/dw_mmc.c
new file:   include/drivers/synopsys/dw_mmc.h
new file:   plat/hisilicon/hikey/include/partitions.h
```
Commented out includes for `gpt.h, mbr.h, dw_usb.h and fastboot.h` because I cannot find them anywhere.
And modified `platform_def.h` and `hikey_io_storage.c`

Is there a switch to build atf without FW Update as before?
Why aren't Linaro and ARM interested in supporting HiKey Platform in this official repo? Since one year nothing happend to the [96boards fork](https://github.com/96boards/arm-trusted-firmware)?

**Comment by danh-arm** [comment]
Hi @gitfineon. We (ARM) are certainly interested in this platform being upstreamed, but it is not in our remit to resolve the outstanding issues with this PR. I agree it would be good to upstream a partial port and fix the bigger issues later.

**Comment by danh-arm** [comment]
I'm closing this PR as it's clearly stalled

**Pull request closed without merging**